staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1019 @@
+	// status.ifcName of an attachment.
+	attachmentIfcIdxName = "ifcName"
+)
+
+// vnState stores all the state needed for a Virtual Network for
+// which there is at least one NetworkAttachment local to this node.
+type vnState struct {
+	// mutex is used to access the map localAtts
+	// and remoteAttsInformer when starting it.
+	mutex sync.RWMutex
+
+	// remoteAttsInformer is an informer on the NetworkAttachments that are
+	// both: (1) in the Virtual Network the vnState represents, (2) not on
+	// this node. It is stopped when the last local NetworkAttachment in the
+	// Virtual Network associated with the vnState instance is deleted. To
+	// stop it, remoteAttsInformerStopCh must be closed.
+	remoteAttsInformer       k8scache.SharedIndexInformer
+	remoteAttsInformerStopCh chan struct{}
+
+	// remoteAttsLister is a lister on the NetworkAttachments that are
+	// both: (1) in the Virtual Network the vnState represents, (2) not
+	// on this node. Since a Virtual Network cannot span multiple k8s API
+	// namespaces, it's a NamespaceLister.
+	remoteAttsLister koslisterv1a1.NetworkAttachmentNamespaceLister
+
+	// localAtts stores the namespaced names the local NetworkAttachments in the Virtual
+	// Network the vnState represents. A local attachment is added to it when
+	// it is seen for the first time, and removed from it when it is deleted
+	// from the local attachments informer cache. Upon being removed, a check
+	// on whether the attachment is the last one in localAtts is made, and if that's
+	// the case remoteAttsInformer is stopped by closing remoteAttsInformerStopCh,
+	// and vnState is removed from the map storing all the vnStates localAtts is used
+	// as a set where the elements are the keys, the map values can be ignored.
+	localAtts map[k8stypes.NamespacedName]struct{}
+}
+
+// vniAndNsn is used as the type for the keys in the maps
+// storing Network Interfaces of the NetworkAttachments.
+// Using a Namespaced name only as the key is not enough
+// because when you have updates to the vni field two different
+// references are enqueued (one for the attachment version with the
+// old vni and one for the attachment version with the new vni).
+// In such cases there is the risk of race conditions if the key
+// is represented by a namespaced name only.
+type vniAndNsn struct {
+	vni uint32
+	nsn k8stypes.NamespacedName
+}
+
+// attQueueRef is a queue reference to a Network Attachment. Upon
+// dequeuing a reference, a worker goroutine can lookup for the
+// attachment in different caches: the local Attachments cache, or
+// one of the caches for remote attachments (one for each Virtual
+// Network relevant to the Connection Agent). That's why attQueueRef
+// stores a locality flag (true if lookup for the attachment must be
+// done in the local attachments cache) and a vni, which identifies
+// the remote attachments cache where to lookup if the locality flag
+// is false.
+type attQueueRef struct {
+	vni   uint32
+	local bool
```

Comment:
To enable a successful lookup it is not necessary to include the `local bool` in a queue entry.  For any VNI & name there are exactly two informers where the corresponding object can be looked up.  Normally the object will be found in at most one of them.  If it is found in both then this is a transient that will resolve with a retry.

Note that changing only the host of a NetworkAttachment does not change its IP or MAC addresses; we probably do not want both the local and remote implementations to exist at any one time.

Reply:
Changes implemented in 
dfcd8d4.